src/sql/platform/backup/common/backupServiceImp.ts

```diff
@@ -27,15 +27,15 @@ export class BackupService implements IBackupService {
 	/**
 	 * Get database metadata needed to populate backup UI
 	 */
-	public getBackupConfigInfo(connectionUri: string): Thenable<azdata.BackupConfigInfo> {
+	public getBackupConfigInfo(connectionUri: string): Promise<azdata.BackupConfigInfo | void> {
```

Comment:
>void [](start = 86, length = 4)

void here seems odd - why did you change it from undefined? It seems like it's more correct to have this return some value all the time since the caller presumably wanted the actual value when they called it

Reply:
Technically void is the absent of a type, where as undefined and null are actually types. so something like Promise.resolve(undefined) would be undefined, and Promise.resolve() would be void. Its more pedantics than anything. https://www.typescriptlang.org/docs/handbook/basic-types.html#void

Reply:
Now in reality javascript treats lack of a value as undefined. Personally when it comes to function I like seeing void if the function isn't going to return anything. It makes more sense when you use async function.

````typescript
async function foo(): Promise<string | void> {
     if(something) {
           return 'some string';
     }
}
````

looks better than me to
````typescript
async function foo(): Promise<string | undefined> {
     if(something) {
           return 'some string';
     }
     return undefined;
}

but they would do the same thing functionally.

Reply:
Although thinking about it when writing a normal function we generally do

````typescript
function foo(): string | undefined {
     if(something) {
           return 'some string';
     }
     return undefined;
}
````

so idk.

Reply:
Maybe the rule should be if you don't ever return anything use `void` if you somethings return anything use `something|undefined` so in the promise land, if your promise never has a value use `Promise<void>`. If your promise sometimes returns something use `Promise<something| undefined>`

Reply:
But like i said javascript functionality will consider `void` like typescript types as undefined so we are really pulling straws at this point.

Reply:
Yeah - so I guess I wasn't sure the behavior of the return value from a void function. Coming from a managed world like C# though string | undefined feels much more natural to me. I also feel like that's the style we tend to use throughout the codebase currently (undefined if no/invalid value).

In the end though I'm fine adopting this if that's agreed to be the way going forward - I just want us to be consistent with usage if we can.

Reply:
Yeah that would be my general preference. `void` only if the function doesn't ever return anything, otherwise `type | undefined` if it does at least some of the time.

Reply:
We should also be looking at using nulls where it makes sense too. We're using null interchangeably with undefined even though they're not entirely the same thing.

But I guess since the typescript team doesn't use null themselves they probably have a good reason.

Reply:
What advantages do you see to using null? We've been moving away from null for a while now and I don't really see any benefit in using that over undefined in nearly all cases.

Reply:
Technically they (pedantically) 2 different things, null is a empty value (like null pointer) where as undefined is a lack of any value. In reality we aren't dealing in areas where that difference matters. The one place we have to care is when interacting with the DOM which something like `element.style.color = undefined` will throw an error in strict null checks because the value is defined in typescript as `string | null`.